server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -1103,25 +1119,29 @@ public void markShardCopyAsStaleIfNeeded(ShardId shardId, String allocationId, l
         private final R request;
         // Indicates if this primary shard request originated by a reroute on this local node.
         private final boolean sentFromLocalReroute;
+        private final boolean localRerouteInitiatedByNodeClient;
 
         public ConcreteShardRequest(Writeable.Reader<R> requestReader, StreamInput in) throws IOException {
             targetAllocationID = in.readString();
             primaryTerm  = in.readVLong();
             sentFromLocalReroute = false;
+            localRerouteInitiatedByNodeClient = false;
```

Comment:
can we also assert (similar  to sentFromLocalReroute) that localRerouteInitiatedByNodeClient is never false when being serialized in writeTo?